### PR TITLE
Reduce the chances of bots reporting analytics

### PR DESCRIPTION
**Pull request by jcollins-g**
Hopefully we'll actually get them all this time.  This fixes the known remaining causes and potential causes of #13843: 

- [Explicit opt-in for analytics allows bots to send data](https://github.com/flutter/flutter/issues/13843#issuecomment-355120336)
- [Flutter repo tests set explicit opt-in of analytics](https://github.com/flutter/flutter/issues/13843#issuecomment-355423320)
- [Bot detection doesn't detect AWS CodeBuild or Jenkins](https://github.com/flutter/flutter/issues/13843#issuecomment-355150789)

### CODE REVIEW

**Review by devoncarew — APPROVED** [review]
Thanks for tracking this down!

File: packages/flutter_tools/lib/src/usage.dart [hunk 1]
```diff
@@ -19,11 +20,14 @@ const String _kFlutterUA = 'UA-67589403-6';
...
-  /// Create a new Usage instance; [versionOverride] is used for testing.
-  Usage({ String settingsName: 'flutter', String versionOverride }) {
+  /// Create a new Usage instance; [versionOverride] and [configDirOverride] are
+  /// used for testing.
+  Usage({ String settingsName: 'flutter', String versionOverride, String configDirOverride}) {
     final FlutterVersion flutterVersion = FlutterVersion.instance;
     final String version = versionOverride ?? flutterVersion.getVersionString(whitelistBranchName: true);
-    _analytics = new AnalyticsIO(_kFlutterUA, settingsName, version);
+    _analytics = new AnalyticsIO(_kFlutterUA, settingsName, version,
+        // ignore: argument_type_not_assignable
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: can you add a comment here explaining the ignore?

File: packages/flutter_tools/lib/src/usage.dart [hunk 2]
```diff
@@ -33,19 +37,13 @@ class Usage {
...
-      runningOnCI = true;
-
-    // Check for common CI systems.
-    if (isRunningOnBot)
-      runningOnCI = true;
-
-    // If we think we're running on a CI system, default to not sending analytics.
-    _analytics.analyticsOpt = runningOnCI ? AnalyticsOpt.optIn : AnalyticsOpt.optOut;
+    if (version.endsWith('/unknown') || isRunningOnBot) {
+      // If we think we're running on a CI system, default to not sending analytics.
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: it's not just defaulting any more, it's actively suppressing regardless of configuration. So we should probably update the comment a little.

**Comment by Hixie** [comment]
This is fantastic. Thanks!

LGTM

File: packages/flutter_tools/lib/src/usage.dart (hunk 1 above)
Comment (jcollins-g) [reply packages/flutter_tools/lib/src/usage.dart]: I won't pretend to fully understand the need for this, but will repeat @tvolkert's explanation here when I asked him about this problem.  I can't seem to track down the location of the conditional import in question, which is probably not that surprising considering I'd ordinarily use the analyzer to figure that out...

File: packages/flutter_tools/lib/src/usage.dart (hunk 2 above)
Comment (jcollins-g) [reply packages/flutter_tools/lib/src/usage.dart]: Done

File: packages/flutter_tools/lib/src/usage.dart [hunk 3]
```diff
@@ -19,11 +20,16 @@ const String _kFlutterUA = 'UA-67589403-6';
...
+  /// Create a new Usage instance; [versionOverride] and [configDirOverride] are
+  /// used for testing.
+  Usage({ String settingsName: 'flutter', String versionOverride, String configDirOverride}) {
     final FlutterVersion flutterVersion = FlutterVersion.instance;
     final String version = versionOverride ?? flutterVersion.getVersionString(whitelistBranchName: true);
-    _analytics = new AnalyticsIO(_kFlutterUA, settingsName, version);
+    _analytics = new AnalyticsIO(_kFlutterUA, settingsName, version,
+        // Analyzer doesn't recognize that [Directory] objects match up due to a
+        // conditional import.
+        // ignore: argument_type_not_assignable
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: fyi @pq

Comment (pq) [reply packages/flutter_tools/lib/src/usage.dart]: Thanks for the heads-up.  I spoke to Janice a bit and it sounds like this is a known problem with a known solution (at least to @bwilkerson :)) but that it's not trivial.  Some relevant background can be gathered from: https://github.com/dart-lang/sdk/issues/30143 but be warned: there be dragons! 😄

**Pull request merged**